### PR TITLE
espanso: 0.7.3 -> 2.1.8

### PR DESCRIPTION
**Pull request by pyrox0**
###### Description of changes
This is a major update of espanso, with breaking(but fixable) changes.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dit7ya** [comment]
Can you please write about how to test this package? Is this supposed to work on Wayland as well?

**Comment by theHedgehog0** [comment]
> Can you please write about how to test this package? Is this supposed to work on Wayland as well?

It is supposed to work on wayland. If you use flakes, I would suggest pulling the branch in as an input and making an overlay or just pulling the package out of the set. Then you could have a small config in your $CONFIG/espanso folder, which would be picked up by the service. This should also be able to build on MacOS, though I have not been able to test it.

**Comment by NickCao** [comment]
As part of https://github.com/NixOS/nixpkgs/issues/210452, would you like to also remove the openssl_1_1 pin in all-packages.nix?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -4442,7 +4442,7 @@ with pkgs;
 
   espanso = callPackage ../applications/office/espanso {
     inherit (darwin.apple_sdk.frameworks) AppKit Cocoa Foundation;
-    openssl = openssl_1_1;
+    # openssl = openssl_1_1;
```
Comment (NickCao) [inline pkgs/top-level/all-packages.nix:4445]: Just remove the line, no need to comment it out.

Comment (theHedgehog0) [reply pkgs/top-level/all-packages.nix:4445]: Was commenting out for testing, will amend the commit.

**Comment by NickCao** [comment]
Unable to test this on wayland:
screenshot-2023-01-17-11-28-39
Does it offer an option to build the wayland variant?

**Comment by theHedgehog0** [comment]
Ok, I'm currently running into some problems where `cargo-make`(the tool used by espanso for building, which allows enabling Wayland support) needs to fetch a dependency at build-time, but network access is disabled and we've already fetched deps by the time that step happens. Any ideas on how to fix this?

This is the log I have from the build
[espanso.log](https://github.com/NixOS/nixpkgs/files/10435944/espanso.log)

**Comment by NickCao** [comment]
Would adding `rust-script` to `nativeBuildInputs` help?

**Comment by PedroHLC** [comment]
@theHedgehog0 any update on this?

File: pkgs/applications/office/espanso/default.nix [hunk 2]
```diff
@@ -1,69 +1,89 @@
...
   doCheck = false;
 
+  preBuild = lib.optionalString waylandSupport ''
+    export NO_X11=true
+  '';
+
   postInstall = ''
     wrapProgram $out/bin/espanso \
-      --prefix PATH : ${lib.makeBinPath [ libnotify xclip ]}
+      --prefix PATH : ${lib.makeBinPath [ libnotify xclip wl-clipboard setxkbmap ]}
```
Comment (hugosenari) [inline pkgs/applications/office/espanso/default.nix:79]: I'm not sure but I think you should verify waylandSupport for wl-clipboard

File: pkgs/applications/office/espanso/default.nix [hunk 3]
```diff
@@ -1,69 +1,89 @@
...
 
   nativeBuildInputs = [
     extra-cmake-modules
     pkg-config
     makeWrapper
+    cargo-make
+    wxGTK32
   ];
 
+  NO_X11 = lib.optionals waylandSupport "true";
```
Comment (AndersonTorres) [inline pkgs/applications/office/espanso/default.nix:48]: Shouldn't this be


```suggestion
  env.NO_X11 = lib.boolToString waylandSupport;
```

?

File: pkgs/applications/office/espanso/default.nix [hunk 4]
```diff
@@ -1,69 +1,89 @@
...
 
-  cargoSha256 = "0ba5skn5s6qh0blf6bvivzvqc2l8v488l9n3x98pmf6nygrikfdb";
+  cargoSha256 = "sha256-U2ccF7DM16TtX3Kc4w4iNV4WsswHJ0FpO3+sWCL1Li8=";
 
   nativeBuildInputs = [
     extra-cmake-modules
     pkg-config
     makeWrapper
+    cargo-make
+    wxGTK32
```
Comment (AndersonTorres) [inline pkgs/applications/office/espanso/default.nix:45]: Is this a nativeBuildInput?

Comment (bobvanderlinden) [reply pkgs/applications/office/espanso/default.nix:45]: I've tried building without wxGTK32 as a nativeBuildInput, but the build failed.

<details>

```
       > error: failed to run custom build command for `espanso-modulo v0.1.0 (/build/source/espanso-modulo)`
       >
       > Caused by:
       >   process didn't exit successfully: `/build/source/target/release/build/espanso-modulo-8653a240257a137b/build-script-build` (exit status: 101)
       >   --- stderr
       >   thread 'main' panicked at 'wxWidgets is not installed, as `wx-config` cannot be executed', espanso-modulo/build.rs:467:5
       >   note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/dnn037vxl868icgfyimfwfcb604x1zfy-espanso-2.1.8.drv'.
```
</details>

**Comment by bobvanderlinden** [comment]
It seems `espanso-0.7.3` is currently broken on nixpkgs-unstable.

<details>

```
       > error[E0793]: reference to packed field is unaligned
       >    --> /build/espanso-0.7.3-vendor.tar.gz/tendril/src/tendril.rs:241:20
       >     |
       > 241 |                 if (*header).refcount.decrement() == 1 {
       >     |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >     |
       >     = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
       >     = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)
       >
       > error[E0793]: reference to packed field is unaligned
       >    --> /build/espanso-0.7.3-vendor.tar.gz/tendril/src/tendril.rs:958:9
       >     |
       > 958 |         (*self.header()).refcount.increment();
       >     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >     |
       >     = note: fields of packed structs are not properly aligned, and creating a misaligned reference is undefined behavior (even if that reference is never dereferenced)
       >     = help: copy the field contents to a local variable, or replace the reference with a raw pointer and use `read_unaligned`/`write_unaligned` (loads and stores via `*p` must be properly aligned even when using raw pointers)
       >
       > For more information about this error, try `rustc --explain E0793`.
       > error: could not compile `tendril` due to 2 previous errors
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/99jxamy1ap48s6ispss4pxzq0zi29w05-espanso-0.7.3.drv'.
```

</details>

This branch works perfectly for me. This is what I did to test it. Note that I don't have much experience with espanso, as this was the first time I wanted to give it a try (but got stuck due to the build errors in 0.7.3).

```
$ nix shell .#espanso
$ espanso --version
espanso 2.1.8
$ espanso service start --unmanaged
$ espanso cmd search
```

I was able to add matches to `~/.config/espanso/match/base.yml`, which showed up in the search and expanded in text fields.

This is all under X11. I have not tested this on Wayland. `waylandSupport = false` or `waylandSupport = true` both seemed to work for me.

I've rebased the branch on nixpkgs-unstable and incorporated the suggestions. You can find the branch here: https://github.com/bobvanderlinden/nixpkgs/tree/espanso-update-2.1.8

**Comment by bobvanderlinden** [comment]
Espanso 0.7.3 is failing and it is part of https://github.com/NixOS/nixpkgs/issues/230712.

The changes in this PR are at least an improvement over a failed build, so I'd vouch for merging.

https://github.com/bobvanderlinden/nixpkgs/tree/espanso-update-2.1.8 handles a number of the feedback. Even if the package hasn't been fully worked out with all conventions, it at least fixes the problem of a failing build on master.

**Comment by AndersonTorres** [comment]
How can this be merged?

**Comment by bobvanderlinden** [comment]
> How can this be merged?

The conflicts are trivial to solve. See https://github.com/bobvanderlinden/nixpkgs/commit/8ca5e19b80a2108f53a14836fc5bda22fa05243f for an up-to-date commit of this pr.

**Comment by AndersonTorres** [comment]
Please do it, preferably linking this PR too.

**Comment by theHedgehog0** [comment]
Hey, very sorry about not updating this in a while, I've been very busy, and FOSS just can't be my priority at the moment. I'm glad someone else is working on this, but I'll close this. I'd still like to be added to the maintainers list for the package, if that would be possible!

**Comment by bobvanderlinden** [comment]
Not a problem! I've created a new one #231062 and kept most of your commit intact. The feedback is incorporated and I've rebase it on nixpkgs-unstable and resolved the conflicts / incompatibilities that popped up (Cargo.lock needed to be added to the PR).

Anyone up for a review?

**Pull request closed without merging**